### PR TITLE
feat: Allow customization of the github.event_name check

### LINKED ISSUE #355: Not merging with "pull_request_target"
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3

### Plugin version

_No response_

### Node.js version

16

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

ubuntu-latest

### Description

Dependabot pr's don't merge when using pull_request_target.

This can be fixed by adding "pull_request_target" with "pull_request" as github events to run with in action.yml. 

I have no problem submitting a pull request adding this little detail, just let me know if you want the fix or if this was a design decision.

### Steps to Reproduce

use pull_request_target for merging pull requests action

### Expected Behavior

Should merge pull requests instead of failing silently.

### LINKED ISSUE #497: Receiving 'Warning: Semver bump '' is invalid!' message
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

v3

### Plugin version

unknown

### Node.js version

latest

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

Ubuntu Latest (Github Action Runner)

### Description

Given the following Automerge Job:
image

We're seeing this error come up:
image

I can't really make heads or tails of this one ... any suggestions on how to troubleshoot it?

Here's the debug log:
```
##[debug]Evaluating condition for step: 'Run fastify/github-action-merge-dependabot@v3'
##[debug]Evaluating: success()
##[debug]Evaluating success:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Run fastify/github-action-merge-dependabot@v3
##[debug]Loading inputs
##[debug]Evaluating: github.token
##[debug]Evaluating Index:
##[debug]..Evaluating github:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'token'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Loading env
Run fastify/github-action-merge-dependabot@v3
##[debug]Evaluating condition for step: 'run'
##[debug]Evaluating: (success() && (github.event_name == 'pull_request') && ((github.actor == 'dependabot[bot]') || (inputs.skip-verification == 'true')))
##[debug]Evaluating And:
##[debug]..Evaluating success:
##[debug]..=> true
##[debug]..Evaluating Equal:
##[debug]....Evaluating Index:
##[debug]......Evaluating github:
##[debug]......=> Object
##[debug]......Evaluating String:
##[debug]......=> 'event_name'
##[debug]....=> 'pull_request_target'
##[debug]....Evaluating String:
##[debug]....=> 'pull_request'
##[debug]..=> false
##[debug]=> false
##[debug]Expanded: (true && ('pull_request_target' == 'pull_request') && ((github['actor'] == 'dependabot[bot]') || (inputs['skip-verification'] == 'true')))
##[debug]Result: false
##[debug]Evaluating condition for step: 'run'
##[debug]Evaluating: success()
##[debug]Evaluating success:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: run
##[debug]Loading inputs
##[debug]Evaluating: inputs.github-token
##[debug]Evaluating Index:
##[debug]..Evaluating inputs:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'github-token'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Evaluating: format('const script = require(''{0}/dist/index.js'')
##[debug]await script({{
##[debug]  github,
##[debug]  context,
##[debug]  inputs: {1},
##[debug]  dependabotMetadata: {{
##[debug]    updateType:  ''{2}'',
##[debug]    dependencyType:''{3}'',
##[debug]    dependencyNames: ''{4}'',
##[debug]  }}
##[debug]}})
##[debug]', github.action_path, toJSON(inputs), steps.dependabot-metadata.outputs.update-type, steps.dependabot-metadata.outputs.dependency-type, steps.dependabot-metadata.outputs.dependency-names)
##[debug]Evaluating format:
##[debug]..Evaluating String:
##[debug]..=> 'const script = require(''{0}/dist/index.js'')
##[debug]await script({{
##[debug]  github,
##[debug]  context,
##[debug]  inputs: {1},
##[debug]  dependabotMetadata: {{
##[debug]    updateType:  ''{2}'',
##[debug]    dependencyType:''{3}'',
##[debug]    dependencyNames: ''{4}'',
##[debug]  }}
##[debug]}})
##[debug]'
##[debug]..Evaluating Index:
##[debug]....Evaluating github:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'action_path'
##[debug]..=> '/home/runner/work/_actions/fastify/github-action-merge-dependabot/v3'
##[debug]..Evaluating toJSON:
##[debug]....Evaluating inputs:
##[debug]....=> Object
##[debug]..=> '{
##[debug]  "approve-only": "false",
##[debug]  "merge-method": "squash",
##[debug]  "use-github-auto-merge": "true",
##[debug]  "target": "patch",
##[debug]  "github-token": "***",
##[debug]  "exclude": "",
##[debug]  "merge-comment": "",
##[debug]  "pr-number": "",
##[debug]  "skip-commit-verification": "false",
##[debug]  "skip-verification": "false"
##[debug]}'
##[debug]..Evaluating Index:
##[debug]....Evaluating Index:
##[debug]......Evaluating Index:
##[debug]........Evaluating steps:
##[debug]........=> Object
##[debug]........Evaluating String:
##[debug]........=> 'dependabot-metadata'
##[debug]......=> Object
##[debug]......Evaluating String:
##[debug]......=> 'outputs'
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'update-type'
##[debug]..=> null
##[debug]..Evaluating Index:
##[debug]....Evaluating Index:
##[debug]......Evaluating Index:
##[debug]........Evaluating steps:
##[debug]........=> Object
##[debug]........Evaluating String:
##[debug]........=> 'dependabot-metadata'
##[debug]......=> Object
##[debug]......Evaluating String:
##[debug]......=> 'outputs'
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'dependency-type'
##[debug]..=> null
##[debug]..Evaluating Index:
##[debug]....Evaluating Index:
##[debug]......Evaluating Index:
##[debug]........Evaluating steps:
##[debug]........=> Object
##[debug]........Evaluating String:
##[debug]........=> 'dependabot-metadata'
##[debug]......=> Object
##[debug]......Evaluating String:
##[debug]......=> 'outputs'
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'dependency-names'
##[debug]..=> null
##[debug]=> 'const script = require(''/home/runner/work/_actions/fastify/github-action-merge-dependabot/v3/dist/index.js'')
##[debug]await script({
##[debug]  github,
##[debug]  context,
##[debug]  inputs: {
##[debug]  "approve-only": "false",
##[debug]  "merge-method": "squash",
##[debug]  "use-github-auto-merge": "true",
##[debug]  "target": "patch",
##[debug]  "github-token": "***",
##[debug]  "exclude": "",
##[debug]  "merge-comment": "",
##[debug]  "pr-number": "",
##[debug]  "skip-commit-verification": "false",
##[debug]  "skip-verification": "false"
##[debug]},
##[debug]  dependabotMetadata: {
##[debug]    updateType:  '''',
##[debug]    dependencyType:'''',
##[debug]    dependencyNames: '''',
##[debug]  }
##[debug]})
##[debug]'
##[debug]Result: 'const script = require(''/home/runner/work/_actions/fastify/github-action-merge-dependabot/v3/dist/index.js'')
##[debug]await script({
##[debug]  github,
##[debug]  context,
##[debug]  inputs: {
##[debug]  "approve-only": "false",
##[debug]  "merge-method": "squash",
##[debug]  "use-github-auto-merge": "true",
##[debug]  "target": "patch",
##[debug]  "github-token": "***",
##[debug]  "exclude": "",
##[debug]  "merge-comment": "",
##[debug]  "pr-number": "",
##[debug]  "skip-commit-verification": "false",
##[debug]  "skip-verification": "false"
##[debug]},
##[debug]  dependabotMetadata: {
##[debug]    updateType:  '''',
##[debug]    dependencyType:'''',
##[debug]    dependencyNames: '''',
##[debug]  }
##[debug]})
##[debug]'
##[debug]Evaluating: (runner.debug == '1')
##[debug]Evaluating Equal:
##[debug]..Evaluating Index:
##[debug]....Evaluating runner:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'debug'
##[debug]..=> '1'
##[debug]..Evaluating String:
##[debug]..=> '1'
##[debug]=> true
##[debug]Expanded: ('1' == '1')
##[debug]Result: true
##[debug]Loading env
Run actions/github-script@v6
request {
  method: 'GET',
  baseUrl: 'https://api.github.com',
  headers: {
    accept: 'application/vnd.github.v3+json',
    'user-agent': 'actions/github-script octokit-core.js/3.6.0 Node.js/16.20.1 (linux; x64)'
  },
  mediaType: { format: '', previews: [] },
  request: {
    agent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object: null prototype],
      requests: [Object: null prototype] {},
      sockets: [Object: null prototype] {},
      freeSockets: [Object: null prototype] {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      scheduling: 'lifo',
      maxTotalSockets: Infinity,
      totalSocketCount: 0,
      maxCachedSessions: 100,
      _sessionCache: [Object],
      [Symbol(kCapture)]: false
    },
    hook: [Function: bound bound register]
  },
  url: '/repos/{owner}/{repo}/pulls/{pull_number}/commits',
  owner: 'MYORG',
  repo: 'MY-REPO',
  pull_number: 157
}
GET /repos/MYORG/MY-REPO/pulls/157/commits - 200 in 347ms
Warning: Semver bump '' is invalid!
##[debug]Node Action run completed with exit code 0
##[debug]Set output merge_status = skipped:invalid_semver
##[debug]Set output result = 
##[debug]Finished: run
##[debug]Finishing: Run fastify/github-action-merge-dependabot@v3
```

And here's the commit text:
```
chore(deps): Bump constructs from 10.2.70 to 10.3.0
Bumps [constructs](https://github.com/aws/constructs) from 10.2.70 to 10.3.0.
- [Release notes](https://github.com/aws/constructs/releases)
- [Commits](https://github.com/aws/constructs/compare/v10.2.70...v10.3.0)

---
updated-dependencies:
- dependency-name: constructs
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
```

### Steps to Reproduce

See above

### Expected Behavior

_No response_

### PR DESCRIPTION
**Pull request by diranged**
The motivation of this PR is to allow the use of automatic merging on `pull_request_target` triggered builds, if so desired. See https://github.com/fastify/github-action-merge-dependabot/issues/355#issuecomment-1755478171 and https://github.com/fastify/github-action-merge-dependabot/issues/497 for context.

Closes #355
Closes #497 

#### Checklist

- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by simoneb — CHANGES_REQUESTED** [review]
I'm requesting changes just to prevent this being merged. As per the comment I added, we need to have docs showing how to use it. Because if you use this trigger but then don't check out the right version of the code, you are in bigger troubles as if you didn't use this trigger in the first place

**Comment by dancamma** [comment]
@diranged I tested the PR and it works! 🥳

However, I think we could achieve the same result without creating a new configuration entry. We can simply modify the if statement to accept both values:

`if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target') && (github.actor == 'dependabot[bot]' || inputs.skip-verification == 'true')`

What do you think?

**Comment by simoneb** [comment]
Folks the main point here is how we prevent people from misusing this. If you use pull_request_target, you're potentially merging a PR after testing it against the base branch, which would be more unsafe than not supporting that trigger in the first place.

**Comment by Uzlopak** [comment]
I think it is actually the best solution to have it as a configuration entry and not in the code, as it makes it possible to explicitly allow pull_request_target.

The documentation part still remains as somebody should be aware, that he is going a potential risky path.

**Comment by Uzlopak** [comment]
Actually the skip-trigger-validation made the less sense for me.

**Comment by Uzlopak** [comment]
Imho this PR has only one flaw: It is not possibel to use this action in a workflow which listens on pull_request and pull_request_target events.

**Comment by diranged** [comment]
> Imho this PR has only one flaw: It is not possibel to use this action in a workflow which listens on pull_request and pull_request_target events.

That is technically true - I can't think of a scenario where that happens though... but you are right. Wouldn't putting in `skip-trigger-validation` as an option be a workaround in that scenario?

**Comment by diranged** [comment]
@simoneb Any update on this one?

**Comment by simoneb** [comment]
My personal opinion is that this should not be supported because it leads to misuse and security issues due to that. I don't own this repository so if anybody has a different opinion I'm happy to hear proposals. 

In general, the easiest way to see anything landing in any project is to submit a PR. Note though that whatever implementation it contains, it should be one that is secure by default.

**Comment by diranged** [comment]
@simoneb :) This _is_ the PR that I submitted... which just gives a user the _option_ of making this work.

**Comment by simoneb** [comment]
Fair enough, let's reopen this then. I didn't notice you had included additional documentation. This makes it look much better now. Approving

**Review by simoneb — APPROVED** [review]

**Comment by simoneb** [comment]
I'm going to merge this now, the changes look good to me as they are now because they add the feature by keeping the default behavior secure, and explaining the steps that need to be taken to adopt the alternative behavior while still preserving security.

**Comment by simoneb** [comment]
@diranged this change seems to have broken the action, do you have time to look into it? https://github.com/fastify/github-action-merge-dependabot/actions/runs/7623277429

File: action.yml [hunk 1]
```diff
@@ -40,14 +40,22 @@ inputs:
     type: boolean
     description: 'If true, the action will not validate the user or the commit verification status'
     default: false
+  event-name:
+    type: string
+    description:
+    default: pull_request
```
Comment (climba03003) [inline action.yml:46]: Seems like the issue of missing quotes, it should be `default: 'pull_request'`

Comment (diranged) [reply action.yml:46]: @climba03003 @simoneb Can someone re-run that in debug mode (I don't have permissions) - I don't quite think this is a quoting issue, but I'd like to see how the evaluation of `${{ inputs.event-name }}` goes.

**Comment by diranged** [comment]
image

That warning message is from https://github.com/dependabot/fetch-metadata/blob/924483a3d758e3539d35e512c5ca92c75c7a748d/src/dependabot/verified_commits.ts#L18-L24 - which seems to be doing a sane-check that you can't run the `dependabot/fetch-metadata` action on _main_, but instead it must run against a PR because that's how it's going to get the PR metadata about the dependencies. (it later fails at https://github.com/dependabot/fetch-metadata/blob/924483a3d758e3539d35e512c5ca92c75c7a748d/src/main.ts#L45-L47).

Looking at the last working build (https://github.com/fastify/github-action-merge-dependabot/actions/runs/7614170037) we can see it fails out with `Error: This action must be used in the context of a Pull Request or with a Pull Request number` - but doesn't actually fail the build because it never ran the `dependabot/fetch-metadata` action.
image
image

I need to see how https://github.com/fastify/github-action-merge-dependabot/pull/500/files#diff-1243c5424efaaa19bd8e813c5e6f6da46316e63761421b3e5f5c8ced9a36e6b6R58 evaluates on `main` builds to understand why it failed..

File: action.yml (hunk 1 above)
Comment (simoneb) [reply action.yml:46]: @diranged I tried a few combinations of with-and-without quotes without success, may I ask some help to fix this? The easiest way to go about it is that you create a branch for the action, where you try your fixes, and push it to a branch on your own remote. 

Then you set up a repo to try out the fix where you reference your own version of the action, meaning that assuming that you call the branch "fix-event-name", you'd reference it in the workflow as

```yaml
uses: diranged/github-action-merge-dependabot@fix-event-name
```

**Comment by simoneb** [comment]
Meanwhile I'm reverting this PR to unbreak the main branch. https://github.com/fastify/github-action-merge-dependabot/pull/563

**Comment by simoneb** [comment]
> image
> That warning message is from https://github.com/dependabot/fetch-metadata/blob/924483a3d758e3539d35e512c5ca92c75c7a748d/src/dependabot/verified_commits.ts#L18-L24 - which seems to be doing a sane-check that you can't run the `dependabot/fetch-metadata` action on _main_, but instead it must run against a PR because that's how it's going to get the PR metadata about the dependencies. (it later fails at https://github.com/dependabot/fetch-metadata/blob/924483a3d758e3539d35e512c5ca92c75c7a748d/src/main.ts#L45-L47).
> 
> Looking at the last working build (https://github.com/fastify/github-action-merge-dependabot/actions/runs/7614170037) we can see it fails out with `Error: This action must be used in the context of a Pull Request or with a Pull Request number` - but doesn't actually fail the build because it never ran the `dependabot/fetch-metadata` action. image image
> 
> I need to see how https://github.com/fastify/github-action-merge-dependabot/pull/500/files#diff-1243c5424efaaa19bd8e813c5e6f6da46316e63761421b3e5f5c8ced9a36e6b6R58 evaluates on `main` builds to understand why it failed..

Just making sure you noticed the comment I posted here https://github.com/fastify/github-action-merge-dependabot/pull/500#discussion_r1465188017

**Pull request merged**